core/templates/dev/head/collection_player/CollectionDataService.js

```diff
@@ -0,0 +1,126 @@
+    '$http', 'COLLECTION_DATA_URL', 'UrlInterpolationService',
+    function($http, COLLECTION_DATA_URL, urlInterpolationService) {
+
+  // Maps previously loaded collections to their IDs.
+  var _collectionCache = [];
+
+  // This inserts an extra 'exploration' object into each node. The object
+  // includes the exploration ID and additional meta-information regarding the
+  // exploration that is useful for collection pages (such as the exploration
+  // title).
+  var _populateExplorationsInNodes = function(collection, explorationTitles) {
+    for (var i = 0; i < collection.nodes.length; i++) {
+      var collectionNode = collection.nodes[i];
+      collectionNode.exploration = {
+        'id': collectionNode.exploration_id,
+        'title': explorationTitles[collectionNode.exploration_id]
+      };
+    }
+    return collection;
+  };
+
+  // TODO(bhenning): Figure out what to do about versioning. Should the version
+  // be specified when a collection is fetched?
+  var _fetchCollection = function(
+      collectionId, successCallback, errorCallback) {
+    var collectionDataUrl = urlInterpolationService.interpolateUrl(
+      COLLECTION_DATA_URL, {'collection_id': collectionId});
+
+    $http.get(collectionDataUrl).success(function(data) {
+      var collection = _populateExplorationsInNodes(
+        angular.copy(data.collection), data.exploration_titles);
+      collection.next_exploration_ids = data.next_exploration_ids;
+      if (successCallback) {
+        successCallback(collection);
+      }
+    }).error(function(data) {
+      if (errorCallback) {
+        errorCallback(data ? data.error : null, collectionId);
+      }
+    });
+  };
+
+  var _isCached = function(collectionId) {
+    return _collectionCache.hasOwnProperty(collectionId);
+  };
+
+  return {
+    /**
+     * Retrieves a collection from the backend given a collection ID. If the
+     * collection is successfully loaded and a successCallback function is
+     * provided, successCallback is called with the collection passed in as a
+     * parameter. If something goes wrong while trying to fetch the collection,
+     * errorCallback is called instead, if present. errorCallback is passed
+     * the error that occurred and the collection ID.
+     */
+    fetchCollection: function(collectionId, successCallback, errorCallback) {
+      _fetchCollection(collectionId, successCallback, errorCallback);
+    },
+
+    /**
+     * Behaves in the exact same way as fetchCollection (including callback
+     * behavior), except this function will attempt to see whether the given
+     * collection has already been loaded. If it has not yet been loaded, it
+     * will fetch the collection from the backend. If it successfully retrieves
+     * the collection from the backend, it will store it in the cache to avoid
+     * requests from the backend in further function calls.
+     *
+     * Note: successCallback may be called before this function returns in the
```

Comment:
This could happen in any case, though, right? I don't think the return is blocking on successCallback even in the case where the collection needs to be fetched from the backend, since $http.get is non-blocking.